controller/volume_attachment_controller.go

```diff
@@ -436,6 +436,26 @@ func (vac *VolumeAttachmentController) handleVolumeMigrationConfirmation(va *lon
 		return
 	}
 
+	// This is not technically a confirmation. However, there is no good reason to allow the volume to remain attached
+	// under these conditions.
+	// - Kubernetes does not want it attached to the old node.
+	// - We can't get a migration engine up on the new down node.
+	// Stop the migration and detach from the old node. If the upper layer changes nothing and the new node comes back,
+	// the volume can exclusively attach to it.
+	if downOrDeleted, err := vac.ds.IsNodeDownOrDeleted(vol.Status.CurrentMigrationNodeID); err != nil {
+		log.WithError(err).Warn("Failed to check if node is down")
+	} else if downOrDeleted {
+		vol.Spec.NodeID = ""
+		vol.Spec.MigrationNodeID = ""
```

Comment:
I was thinking the migration would be confirmed as usual, and the volume would become detached when the new engine (the related IM pod) becomes terminating. 
In other words, should this controller be responsible for detaching the volumes? Can we still rely on the engine crash logic inside VolumeController?

Reply:
Good suggestion. I tried this out with the following changes (essentially allowing a normal migration confirmation instead of a detachment):

```
        if downOrDeleted, err := vac.ds.IsNodeDownOrDeleted(vol.Status.CurrentMigrationNodeID); err != nil {
                log.WithError(err).Warn("Failed to check if node is down")
        } else if downOrDeleted {
-               vol.Spec.NodeID = ""
+               vol.Spec.NodeID = vol.Status.CurrentMigrationNodeID
                vol.Spec.MigrationNodeID = ""
                log = getLoggerForMigratingLHVolumeAttachment(vac.logger, va, vol)
-               log.Warn("Detaching volume for attempted migration to down node")
+               log.Warn("Doing Shuo's suggested behavior")
                return
        }
```

In the couple of tests I did, everything worked fine. The `currentNodeID` switched as expected and the usual reattachment logic started trying to reattach to the new node (which was down).

However, this behavior makes me nervous. In the past, we have never allowed a migration confirmation when the new engine is not available. I think doing so leaves us open to potential races that I have not accounted for.

> In other words, should this controller be responsible for detaching the volumes?

This is the only controller that is responsibly for detaching volumes. While the usual logic that sets `vol.Spec.NodeID = ""` is in a different function, it is a different function within this controller.

I think it is safest to maintain the already implemented/tested behavior. In practice, it is very similar to your suggestion, but it does not switch the active engine and replicas (which is the part that is a bit dicey when the migration engine is actually not running IMO).

Reply:
Discussed this with @shuo-wu offline and we are fine to proceed with the current implementation.